radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,746 @@
+                    id,
+                    Redactable::Present(Revision::new(author, current, proposed, timestamp)),
+                ),
+                Action::Thread { revision, action } => match self.revisions.get_mut(&revision) {
+                    Some(Redactable::Present(revision)) => revision
+                        .discussion
+                        .apply([cob::Op::new(action, op.author, op.timestamp, op.clock)])?,
+                    Some(Redactable::Redacted) => return Err(ApplyError::Redacted(revision)),
+                    None => return Err(ApplyError::Missing(revision)),
+                },
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Verdict {
+    /// An accepting verdict must supply the [`Signature`] over the
+    /// new proposed [`Doc`].
+    Accept(Signature),
+    /// Rejecting the proposed [`Doc`].
+    Reject,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Revision {
+    /// Author of this proposed revision.
+    pub author: Author,
+    /// [`Identity::current`]'s current [`Oid`] that this revision was
+    /// based on.
+    pub current: Oid,
+    /// New [`Doc`] that will replace `previous`' document.
+    pub proposed: Doc<Verified>,
+    /// Discussion thread for this revision.
+    pub discussion: Thread,
+    /// [`Verdict`]s given by the delegates.
+    pub verdicts: LWWMap<PublicKey, Redactable<Verdict>>,
+    /// Physical timestamp of this proposal revision.
+    pub timestamp: Timestamp,
+}
+
+impl Revision {
+    pub fn new(
+        author: Author,
+        current: Oid,
+        proposed: Doc<Verified>,
+        timestamp: Timestamp,
+    ) -> Self {
+        Self {
+            author,
+            current,
+            proposed,
+            discussion: Thread::default(),
+            verdicts: LWWMap::default(),
+            timestamp,
+        }
+    }
+
+    pub fn signatures(&self) -> impl Iterator<Item = (&PublicKey, Signature)> {
+        self.verdicts().filter_map(|(key, verdict)| match verdict {
+            Verdict::Accept(sig) => Some((key, *sig)),
+            Verdict::Reject => None,
+        })
+    }
+
+    pub fn verdicts(&self) -> impl Iterator<Item = (&PublicKey, &Verdict)> {
+        self.verdicts
+            .iter()
+            .filter_map(|(key, verdict)| verdict.get().map(|verdict| (key, verdict)))
+    }
+
+    pub fn accepted(&self) -> Vec<PublicKey> {
+        self.verdicts()
+            .filter_map(|(key, v)| match v {
+                Verdict::Accept(_) => Some(*key),
+                Verdict::Reject => None,
+            })
+            .collect()
+    }
+
+    pub fn rejected(&self) -> Vec<PublicKey> {
+        self.verdicts()
+            .filter_map(|(key, v)| match v {
+                Verdict::Accept(_) => None,
+                Verdict::Reject => Some(*key),
+            })
+            .collect()
+    }
+
+    pub fn reaches_quorum(&self, previous: &Identity<Oid>) -> bool {
```

Comment:
We have this convention for predicate functions to use `is_*` which I'd like to keep, so maybe `is_quorum_reached`?